In [5]:
# --- Бібліотеки для даної роботи ---
try:
    import numpy, pandas, scipy, matplotlib, jinja2
    print("Бібліотеки вже встановлені. Пропускаємо інсталяцію.")
except ImportError:
    print("Встановлюємо бібліотеки...")
    %pip install numpy pandas scipy matplotlib jinja2

Бібліотеки вже встановлені. Пропускаємо інсталяцію.


## Задача 3. Аналіз ефективності рекламної кампанії

**Умова:**

Ви працюєте аналітиком в EdTech-компанії, яка запустила новий онлайн-курс. Маркетинговий відділ хоче оцінити загальну кількість реєстрацій за перший тиждень та спрогнозувати теоретичний максимум охоплення.

На основі даних попередніх запусків встановлено, що інтенсивність реєстрацій (кількість нових реєстрацій за день) описується функцією:
$$f(t) = 500 \cdot e^{-0.3t}$$
де $t$ — кількість днів після запуску, а $f(t)$ — кількість реєстрацій на день.

Функція показує, що найбільший інтерес до курсу в перші дні після запуску, потім він поступово згасає.

**Виконайте наступні завдання:**

1.  **Обчисліть аналітично кількість реєстрацій за перші 7 днів.** Знайдіть визначений інтеграл $\int_{0}^{7} 500 \cdot e^{-0.3t} dt$ за формулою Ньютона-Лейбніца. Підказка: первісна для функції $e^{at}$ має вигляд $F(t) = \frac{1}{a}e^{at} + C$. Наприклад, $\int e^{-0.3t} dt = \frac{1}{-0.3}e^{-0.3t} + C = -\frac{10}{3}e^{-0.3t} + C.$
2.  **Перевірте результат чисельно.** Використовуючи функцію `quad` з бібліотеки `scipy.integrate`, обчисліть той самий інтеграл чисельно. Порівняйте з аналітичним результатом.
3.  **Обчисліть теоретичний максимум реєстрацій.** Знайдіть невласний інтеграл $\int_{0}^{\infty} 500 \cdot e^{-0.3t} dt$ — це загальна кількість реєстрацій, якби кампанія тривала нескінченно довго. Підказка: при $t \to \infty$ значення $e^{-0.3t} \to 0$.
4.  **Визначте ефективність першого тижня.** Обчисліть, який відсоток від теоретичного максимуму становлять реєстрації за перші 7 днів.

In [6]:
import numpy as np
import pandas as pd
from scipy.integrate import quad
from IPython.display import display

def registration_rate(t):
    return 500 * np.exp(-0.3 * t)

integral_7_days, error_7 = quad(registration_rate, 0, 7)

integral_infinity, error_inf = quad(registration_rate, 0, np.inf)

efficiency_percent = (integral_7_days / integral_infinity) * 100

def analytical_integral_value(t):
    return -(500 / 0.3) * np.exp(-0.3 * t)

analytical_7 = analytical_integral_value(7) - analytical_integral_value(0)
analytical_inf = 0 - analytical_integral_value(0) 

print("Красивий Вивід:")
print("-" * 60)
print("ЗВІТ ЕФЕКТИВНОСТІ ЗАПУСКУ КУРСУ")
print("-" * 60)

data = [
    {"Показник": "Реєстрації за 1-й тиждень (Чисельно)", "Значення": integral_7_days, "Од.": "користувачів"},
    {"Показник": "Реєстрації за 1-й тиждень (Аналітично)", "Значення": analytical_7, "Од.": "користувачів"},
    {"Показник": "Теоретичний максимум (t -> ∞)", "Значення": integral_infinity, "Од.": "користувачів"},
    {"Показник": "Ефективність першого тижня", "Значення": efficiency_percent, "Од.": "%"}
]

df = pd.DataFrame(data)

styled_df = df.style.format({
    "Значення": "{:.2f}"
}).hide(axis='index')

display(styled_df)

print(f"\nВИСНОВОК: За перші 7 днів ми отримали {efficiency_percent:.1f}% від усіх можливих лідів.")
print("Кампанія дуже динамічна, основний трафік приходить на початку.")
print("=" * 60)

print("\nТехнічний вивід:")
technical_output = {
    "інтеграл_0_7_чисельно": integral_7_days,
    "інтеграл_0_7_аналітично": analytical_7,
    "інтеграл_0_неск_чисельно": integral_infinity,
    "інтеграл_0_неск_аналітично": analytical_inf,
    "абсолютна_похибка": abs(integral_7_days - analytical_7),
    "коефіцієнт_ефективності": efficiency_percent / 100
}
print(technical_output)

Красивий Вивід:
------------------------------------------------------------
ЗВІТ ЕФЕКТИВНОСТІ ЗАПУСКУ КУРСУ
------------------------------------------------------------



ВИСНОВОК: За перші 7 днів ми отримали 87.8% від усіх можливих лідів.
Кампанія дуже динамічна, основний трафік приходить на початку.

Технічний вивід:
{'інтеграл_0_7_чисельно': 1462.5726195783636, 'інтеграл_0_7_аналітично': np.float64(1462.5726195783636), 'інтеграл_0_неск_чисельно': 1666.6666666666663, 'інтеграл_0_неск_аналітично': np.float64(1666.6666666666667), 'абсолютна_похибка': np.float64(0.0), 'коефіцієнт_ефективності': 0.8775435717470184}


### Аналітичні обчислення

#### 1. Кількість реєстрацій за 7 днів (Визначений інтеграл)

Нам потрібно обчислити:
$$I_7 = \int_{0}^{7} 500 \cdot e^{-0.3t} \, dt$$

Виносимо сталу за знак інтеграла:
$$I_7 = 500 \cdot \int_{0}^{7} e^{-0.3t} \, dt$$

Знаходимо первісну $F(t)$. Використовуємо формулу $\int e^{at} dt = \frac{1}{a}e^{at}$:
$$F(t) = 500 \cdot \left( \frac{1}{-0.3} e^{-0.3t} \right) = -\frac{500}{0.3} e^{-0.3t} \approx -1666.67 \cdot e^{-0.3t}$$

Застосовуємо формулу Ньютона-Лейбніца $F(b) - F(a)$:
$$I_7 = \left[ -\frac{500}{0.3} e^{-0.3 \cdot 7} \right] - \left[ -\frac{500}{0.3} e^{-0.3 \cdot 0} \right]$$

$$I_7 = -\frac{500}{0.3} (e^{-2.1} - e^{0})$$

Підставляємо значення ($e^0 = 1$, $e^{-2.1} \approx 0.122456$):
$$I_7 = -1666.666... \cdot (0.122456 - 1)$$
$$I_7 = -1666.666... \cdot (-0.877544)$$
$$I_7 \approx 1462.57$$

**Відповідь:** За перші 7 днів очікується **1463** реєстрації.

#### 2. Теоретичний максимум (Невласний інтеграл)

Обчислюємо загальну кількість за весь час ($t \to \infty$):
$$I_{\infty} = \int_{0}^{\infty} 500 \cdot e^{-0.3t} \, dt$$

$$I_{\infty} = \lim_{b \to \infty} \left( -\frac{500}{0.3} e^{-0.3b} \right) - \left( -\frac{500}{0.3} e^{0} \right)$$

Оскільки $\lim_{b \to \infty} e^{-0.3b} = 0$:
$$I_{\infty} = 0 - (-1666.666...) = \frac{500}{0.3} \approx 1666.67$$

**Відповідь:** Теоретичний максимум становить **1667** реєстрацій.

#### 3. Ефективність першого тижня

Знаходимо відношення отриманих реєстрацій до максимуму:
$$\text{Efficiency} = \frac{I_7}{I_{\infty}} \cdot 100\%$$

Зауважимо, що $I_7 = I_{\infty} \cdot (1 - e^{-2.1})$, тому відношення дорівнює просто $(1 - e^{-2.1}) \cdot 100\%$:
$$\text{Efficiency} = (1 - 0.122456) \cdot 100\% = 87.7544\%$$

**Відповідь:** За перший тиждень отримано **87.75%** від усіх можливих реєстрацій.